File: framework/go/src/unittest/validate.go
package unittest

import (
    "reflect"
)

type SliceValidateFunc func(a, b interface{}) bool

type ValidatorsType struct{}

var Validators ValidatorsType

func (v *ValidatorsType) ForSlice(objType reflect.Type, ordered bool) SliceValidateFunc {
    return v.forSlice(ordered, FeatureFactory.Create(objType))
}

func (v *ValidatorsType) ForSlice2d(objType reflect.Type, dim1Ordered, dim2Ordered bool) SliceValidateFunc {
    elemFeat := FeatureFactory.Create(objType)
    var f ObjectFeature
    if dim2Ordered {
        f = ListFeature{ElemFeat: elemFeat}
    } else {
        f = UnorderedListFeature{ElemFeat: elemFeat}
    }
    return v.forSlice(dim1Ordered, f)
}

func (v *ValidatorsType) forSlice(ordered bool, elemFeat ObjectFeature) SliceValidateFunc {
    if ordered {
        return ListFeature{ElemFeat: elemFeat}.IsEqual
    }
    return UnorderedListFeature{ElemFeat: elemFeat}.IsEqual
}

File: framework/python/src/soda/unittest/validate.py
from typing import *

from .featurefactory import (
    FeatureFactory,
    ListFeature,
    UnorderedListFeature
)

class Validators:

    @classmethod
    def forList(cls, objType: type, ordered: bool) -> Callable[[Any,Any],bool]:
        d = FeatureFactory.create(objType)
        return ListFeature(d).isEqual if ordered else UnorderedListFeature(d).isEqual

    @classmethod
    def forList2d(cls, objType: type, dim1Ordered: bool, dim2Ordered: bool) -> Callable[[Any,Any],bool]:
        elemFeat = FeatureFactory.create(objType)
        d = ListFeature(elemFeat) if dim2Ordered else UnorderedListFeature(elemFeat)
        return ListFeature(d).isEqual if dim1Ordered else UnorderedListFeature(d).isEqual
